Step4 코드 리뷰 드립니다.

lcjltj opened the pull request:
마지막 미션인데 굉장히.. 어렵네요.

리뷰어님 덕분에 정말 많은걸 배우고 있는데..

뒤로 갈수록 코드를 지저분하게 짜는거같아 .. 기능은 우선 개발한거 같아 리뷰 한번 요청드릴게요 ㅠㅠ

lcjltj committed:
feat: Line Entity Charge 컬럼 추가 외 1

- Line: 노선 별 기본 요금 컬럼 추가
- Charge: 요금 관련 VO 추가 및 테스트 케이스 작성

lcjltj committed:
refactor: Line Entity Charge 컬럼 변경

- Line: charge -> extraCharge 추가 요금 column으로 변경
- LineRequest: extraCharge 필드 추가

lcjltj committed:
refactor: PathFinder 최단 거리에서 사용된 라인의 최대 추가 금액 추가

- JgraphPathFinder: 최단 거리에서 사용된 노선 중 추가 금액이 있는 라인을 구하기 위해
                   SectionWeightedEdge 추가

lcjltj committed:
refactor: 노선 ExtraCharge column 수정

- Line: 노선에 있는 ExtraCharge column은 추가 요금에 대한 관심만 가지고
  있어야 기본 요금이 변경 되어도 변경을 최소화 할수 있음으로 추가 요금에 대한
  관심만 가지고 있도록 수정.

lcjltj committed:
refactor: AuthenticationPrincipal required 메소드 추가

- AuthenticationPrincipalArgumentResolver: 비회원 요청 일   경우 (required = false / accessToken = null) NonMember return
- JgraphPathFinder: 최단 거리 조회 후 이용 한 노선 중 추가 요금이 가장 비싼 요금 계산

lcjltj committed:
feat: 거리 요금 정책 / 나이 요금 정책 추가

lcjltj committed:
refactor: DistanceFarePolicy 리팩토링(상수 적용, 거리 계산 방식 변경)

ohtaeg commented on src/test/java/nextstep/subway/favorite/FavoriteAcceptanceTest.java:
```diff
@@ -50,9 +50,9 @@ public void setUp() {
         교대역 = StationAcceptanceTest.지하철역_등록되어_있음("교대역").as(StationResponse.class);
         남부터미널역 = StationAcceptanceTest.지하철역_등록되어_있음("남부터미널역").as(StationResponse.class);
 
```
즐겨찾기 관리 테스트가 실패하고 있는 것 같아요 😭 
한번 더 확인해주시면 감사하겠습니다 :)

lcjltj replied:
![image](https://user-images.githubusercontent.com/24871928/176446946-a1599e2e-01e6-4317-937d-c437496b9c88.png)
ㅠ_ㅠ 푸쉬하기 전에 항상 테스트케이스 모두가 정상인지 확인하고 푸쉬를 하고 있는데ㅠ ㅠㅠㅠㅠ
제 PC에서는 테스트케이스가 다 성공하고있어서요 ㅠㅠ

lcjltj replied:
수정 해씁니다.!  ㅠㅠ .. 노선 추가요금이랑 나이를 잘못 하고있엇네요 ㅠㅠㅠㅠㅠ

ohtaeg commented on src/main/java/nextstep/subway/auth/domain/LoginMember.java:
```diff
@@ -6,6 +6,8 @@
 import java.util.Objects;
 
 public class LoginMember {
+    private static final LoginMember NON_MEMBER = new LoginMember(0L, "none@member.com", 20);
```
👍

ohtaeg commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -25,7 +26,13 @@ public boolean supportsParameter(MethodParameter parameter) {
     @Override
     public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
-        String credentials = AuthorizationExtractor.extract(webRequest.getNativeRequest(HttpServletRequest.class));
+        final String credentials = AuthorizationExtractor.extract(webRequest.getNativeRequest(HttpServletRequest.class));
+        final boolean required = parameter.getParameterAnnotation(AuthenticationPrincipal.class).required();
+
+        if (!required && credentials == null) {
+            return LoginMember.ofNonMember();
+        }
```
👍 👍

ohtaeg commented on src/main/java/nextstep/subway/fare/domain/AgeFarePolicy.java:
```diff
@@ -0,0 +1,46 @@
+    INFANT(0, 6, fare -> 0),
+    KIDS(6, 13, fare -> (fare - 350) * 5 / 10),
+    TEENAGER(13, 19, fare -> (fare - 350) * 8 / 10),
+    ADULT(20, Integer.MAX_VALUE, fare -> fare);
```
21억살 👍

ohtaeg commented on src/main/java/nextstep/subway/fare/domain/AgeFarePolicy.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final int beginAge;
+    private final int endAge;
+    private final Function<Integer, Integer> discount;
```
연령대에 맞는 람다식 👍

ohtaeg commented on src/main/java/nextstep/subway/fare/domain/Fare.java:
```diff
@@ -0,0 +1,33 @@
+
+    public static Fare of(final LoginMember loginMember, final int lineExtraCharge, final int distance) {
+        final int totalCharge = totalCharge(lineExtraCharge, distance);
+        if (isNonMember(loginMember)) {
```
별도로 아래에서 메서드를 만들지말고 바로 호출해도 되지 않을까해요 :D
```suggestion
        if (loginMember.isNonMember()) {
```

ohtaeg commented on src/main/java/nextstep/subway/path/dto/PathResponse.java:
```diff
@@ -1,24 +1,27 @@
+    private final int fare;
 
-    public PathResponse(final List<Station> stations, final int distance) {
+    public PathResponse(final List<Station> stations, final int distance, final Fare fare) {
```
요금 계산 책임을 잘 분리해주셨어요 💯
PathService에서 최단 경로를 구한 후 요금 계산을 담당하는 Fare에게 계산을 하도록 요청을 보내고 있도록 해주셨네요 👍 👍

ohtaeg commented on src/main/java/nextstep/subway/path/domain/JgraphPathFinder.java:
```diff
@@ -1,29 +1,49 @@
+        return StationPath.of(path.getVertexList(), path.getWeight(), extraCharge);
     }
 
+    private Integer getExtraCharge(final List<SectionWeightedEdge> sectionEdge) {
```
primitive 타입이 아닌 Wrapper 클래스로 반환해주시는 의도가 따로 있으실까요 :) ?

ohtaeg commented on src/main/java/nextstep/subway/path/domain/JgraphPathFinder.java:
```diff
@@ -33,7 +53,7 @@ private void validation(final Station start, final Station destination) {
     private void addStation(final List<Line> lines) {
         for (Line line : lines) {
             line.getSortedStations()
-                    .forEach(it -> graph.addVertex(it));
+                    .forEach(graph::addVertex);
```
이중 포문 발견..! 
개인적으로 솔직하게 실무에서는 가끔 반복문 내에 if문을 작성하곤 합니다 😂
아쉽지만 우테켐 과정에서는 depth 1(반복문 내부에 for문 허용 x)만 허용하기에
해당 부분을 리팩토링 해주시면 좋을 것 같아요 :)

lcjltj replied:
stream 안의 foreach도 dept로 허용하는군요. ㅠㅠ
flatMap을 사용 하여, 처리르 하긴 하였는데. 보통 저는 스트림 안에 포문이 또 들어가게되면 밖의 stream은 포문으로 작성하고 
안의 for문은  stream을 통해 처리하는게 가독성이 좋다고 생각 하여 그렇게 작성하는 편인데 리뷰어님은 어떤식으로 하시는 편인가요!!??

ohtaeg replied:
저도 찬준님이 말씀해주신대로 중첩 포문이 되는 경우라면 for문으로 표현하여 가독성을 높히려구 합니다 :)
추가로 만약 getter를 사용한 stream을 이용하는 경우 Consumer도 고려해보기도 합니다!
```
public void forEach(Consumer<Station> consumer) {
        getSortedStations().forEach(consumer);
}
```

ohtaeg commented on src/main/java/nextstep/subway/path/domain/JgraphPathFinder.java:
```diff
@@ -1,29 +1,49 @@
+    private Integer getExtraCharge(final List<SectionWeightedEdge> sectionEdge) {
+        return sectionEdge.stream()
+                .map(SectionWeightedEdge::getSection)
+                .map(Section::getLine)
+                .map(Line::getExtraCharge)
```
현재 getter 호출이 많이 일어나는 구조같아요!
이 로직을 SectionWeightedEdge에 메서드로 분리하는 것도 좋을 것 같아요 :)

lcjltj replied:
해당 부분을 작성할때.. getter를  너무 많이 사용 하여, 로직이 보기 안좋다고 생각했는데..  생각해보니 리뷰어님 말대로 특정 메소드를 통하는 방식으로 하는편이 코드를  읽는편에도 더 좋을거 같아요

ohtaeg commented on src/test/java/nextstep/subway/fare/domain/DistanceFarePolicyTest.java:
```diff
@@ -0,0 +1,36 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class DistanceFarePolicyTest {
```
추가된 도메인마다 꼼꼼하게 테스트 👍

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -68,6 +68,8 @@ void shortest_path() {
         // then
         assertThat(최단거리_조회됨.getStations()).hasSize(3);
         assertThat(최단거리_조회됨.getDistance()).isEqualTo(20);
+        // 비회원(1,250 + 노선 추가요금(신분당선) 1,000 + 거리(20km) 추가요금 200
+        assertThat(최단거리_조회됨.getFare()).isEqualTo(2450);
```
여러케이스를 테스트해보았는데 잘되네요 👍

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -68,6 +68,8 @@ void shortest_path() {
         // then
         assertThat(최단거리_조회됨.getStations()).hasSize(3);
         assertThat(최단거리_조회됨.getDistance()).isEqualTo(20);
+        // 비회원(1,250 + 노선 추가요금(신분당선) 1,000 + 거리(20km) 추가요금 200
```
회원, 비회원 시나리오가 별도로 분리되어 테스트가 되면 더 좋을 것 같아요..!
두 시나리오 중에 하나는 성인이 아닌 시나리오라면 더 좋을 것 같네요 😄

ohtaeg reviewed (changes requested):
마지막 미션도 구현을 잘해주셨어요 👍 👍 
말씀해주신대로 요구사항을 잘 만족해주셨습니다 👍 
제가 조금이나마라도 찬준님께 도움을 드린 것 같아서 다행이네요 😂 🙇 

소소한 코멘트를 드렸는데 해당 코멘트들이 반영이 된다면 가꾸셨던 열매를 잘 맺고 가실 수 있을 것 같아서
request changes 하였습니다 😄  
마지막 피드백이 될 것 같네요 :) 마지막까지 화이팅입니다 🔥

lcjltj committed:
refactor: feedback 반영

- JgraphPathFinder: addStation() 이중 포문 제거
- SectionWeightedEdge: getExtraCharge() 메소드 추가
- PathAcceptanceTest: 비 회원/ 회원(청소년) 인수 테스트 추가

lcjltj committed:
fix: 즐겨찾기 인수테스트 수정

lcjltj commented:
안녕하세요. 리뷰어님..
해당 미션이 거의 마무리 단계인지라. 질문이 있어 질문드립니다.
만약 지하철 역의 삭제 되면 그에 영향 있는 즐겨 찾기를 삭제하도록 하고 싶은데. 즐겨 찾기 인 경우에는  source, target 두개 모두 Station을 사용하고 있어, 지하철역에서 mappedBy를 어떻게 준 후 cascade와 orhapRemove를 줘야 하는지 감이 안와서요..
참고할만한 사이트같은게 있을까요?

ohtaeg commented:
음..살펴보았는데, 영한님 책 혹은 인강 외에는 참고할만한 사이트를 알려드리기가 어렵네요 😭 🙏 
지하철역이 삭제가 된다면 영향있는 즐겨찾기들은 아래와 같은 방법을 한번 이용해보시는 것도 괜찮을 것 같아요 :)
`@OnDelete(action = OnDeleteAction.CASCADE)`

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -46,31 +51,61 @@ public void setUp() {
     }
 
-    public static ExtractableResponse<Response> 지하철_최단거리_조회(final Long source, final Long target) {
+    @Test
+    @DisplayName("청소년 최단 거리 조회")
+    void teenagerShortestPath() {
```
피드백 반영 👍 👍

ohtaeg commented on src/main/java/nextstep/subway/path/domain/JgraphPathFinder.java:
```diff
@@ -1,40 +1,57 @@
-            line.getSortedStations()
-                    .forEach(it -> graph.addVertex(it));
-        }
+        lines.stream()
+                .flatMap(it -> it.getSortedStations().stream())
+                .forEach(graph::addVertex);
```
> stream 안의 foreach도 dept로 허용하는군요. ㅠㅠ
flatMap을 사용 하여, 처리르 하긴 하였는데. 보통 저는 스트림 안에 포문이 또 들어가게되면 밖의 stream은 포문으로 작성하고
안의 for문은 stream을 통해 처리하는게 가독성이 좋다고 생각 하여 그렇게 작성하는 편인데 리뷰어님은 어떤식으로 하시는 편인가요!!??

저도 찬준님이 말씀해주신대로 중첩 포문이 되는 경우라면 for문으로 표현하여 가독성을 높히려구 합니다 :)
추가로 만약 getter를 사용하여 컬렉션을 갖고와서 stream을 이용하는 경우 이미 알고있는 의존성인지 등을 고려해서
Consumer도 고려해보기도 합니다 :)

```
// Line.class
public void forEach(Consumer<Station> consumer) {
    getSortedStations().forEach(consumer);
}
```

ohtaeg reviewed (approved):
마지막까지 피드백 반영을 잘해주셨어요 👍 👍 
마지막으로 소소한 코멘트를 드렸는데 확인해주시면 감사하겠습니다 😄 

ATDD + TDD 미션 완주를 축하드립니다 🎉 
많은 피드백들을 반영하시느라 고생많으셨어요 🙇 
조금만 더 힘내셔서 남은 미션들도 화이팅하시길 바라겠습니다 💪
